include/seqan/bam_io/bam_index_bai.h

```diff
@@ -154,6 +154,160 @@ class BamIndex<Bai>
+
+    --end; // zero based index
+
+    list.clear();
+
+    unsigned k;
+    list.push_back(0);
+    for (k =    1 + (beg>>26); k <=    1 + (end>>26); ++k) list.push_back(k);
+    for (k =    9 + (beg>>23); k <=    9 + (end>>23); ++k) list.push_back(k);
+    for (k =   73 + (beg>>20); k <=   73 + (end>>20); ++k) list.push_back(k);
+    for (k =  585 + (beg>>17); k <=  585 + (end>>17); ++k) list.push_back(k);
+    for (k = 4681 + (beg>>14); k <= 4681 + (end>>14); ++k) list.push_back(k);
+}
+
+inline bool
+_getMinFileOffset(uint64_t & minFileOffset,
+                  BamIndex<Bai> const & bamIndex,
+                  int32_t refId,
+                  int32_t position)
+{
+    unsigned windowIdx = position >> 14; // Linear index consists of 16kb windows.
+
+    if (windowIdx >= length(bamIndex._linearIndices[refId]))
+        return false;
+
+    minFileOffset = bamIndex._linearIndices[refId][windowIdx];
+
+    return true;
+}
+
+template <typename TContainer, typename TSpec>
+inline
+SEQAN_FUNC_ENABLE_IF(And<IsSameType<typename Value<TContainer>::Type, BamAlignmentRecord>,
+                     Not<IsSameType<typename Value<TContainer>::Type, TContainer>>>, bool)
+viewRecords(TContainer & resultContainer,
+            FormattedFile<Bam, Input, TSpec> & bamFile,
+            BamIndex<Bai> const & bamIndex,
+            int32_t refId,
+            int32_t regionStart,
+            int32_t regionEnd)
+{
+    typedef typename std::map<uint32_t, BaiBamIndexBinData_>::const_iterator TMapIter;
+    typedef String<Pair<uint64_t, uint64_t>> TChunk;
+    typedef Iterator<TChunk const, Rooted>::Type TChunkIter;
+
+    // Sanity checks
+    // -------------------------------------------------------------------------
+    if (!isEqual(format(bamFile), Bam()))
+        return false;
+
+    if (refId < 0)
+        return false;  // Cannot seek to invalid reference.
+
+    if (static_cast<unsigned>(refId) >= length(bamIndex._binIndices))
+        return false;  // Cannot seek to invalid reference.
+
+    if (regionStart > regionEnd)
+        return false; // invalid region
+
+    if (regionEnd > 536870912)
+        return false; // BAM format only allows reference of size < 2^29
+
+    // Adjust region
+    // -------------------------------------------------------------------------
+    // We want to mimic samtools view. Since SAM/BAM files are one based we expect
+    // the given region to be also one-based (To be consistent with samtools).
+    // Since SeqAn works 0-based we need to substract 1.
+    SEQAN_ASSERT_GEQ_MSG(regionStart, 1, "Expect 1-based region interval which excludes 0 as a start pos.");
+    --regionStart;
+    --regionEnd;
+
+    // Retrieve the candidate bin identifiers for [regionStart, regionEnd).
+    // -------------------------------------------------------------------------
+    std::vector<uint16_t> candidateBins;
+    _baiReg2bins(candidateBins, regionStart, regionEnd);
```

Comment:
Careful, I think in _baiReg2bins you still expect 1-based positions, don't you?
At least for `regionEnd`?

Reply:
No and yes. The function expects 0-based coordinates but was computing a half open interval (reason for `--end` ). Since we want to mimic samtools and samtools computes a closed interval we will do the same now. So thanks for pointing it out.